MG-1987 - Enable Entity Endpoints to Return Basic Info for non-admins.

WashingtonKK opened the pull request:
<!-- Copyright (c) Abstract Machines
SPDX-License-Identifier: Apache-2.0 -->

# What type of PR is this?

This is a feature because it adds the capability for entity endpoints to return basic entity information for non privileged users.

## What does this do?

This PR allows non-admin users to get basic information (name and id) from entities without needing any administrator privileges.

## Which issue(s) does this PR fix/relate to?

- Resolves #1987

## Have you included tests for your changes?

No

## Did you document any new/modified feature?

Yes, I have updated the documentation for the new feature.

### Notes
- This can be used to fetch user names for invitations to allow user names to be acquired without needing admin privileges.

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
feat: add name to domain and user

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Refactor tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

update mocks

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Update mocks

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Get basic domain info

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Update invitation to query names:

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: ci

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix: ci

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Use entity to hold domain and user

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

refinements

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

update retrieve all basic info

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

replace repo call with service filtering

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Update test

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

dborovcanin commented on auth/postgres/domains.go:
```diff
@@ -96,6 +96,36 @@ func (repo domainRepo) RetrieveByID(ctx context.Context, id string) (auth.Domain
 	return auth.Domain{}, repoerr.ErrNotFound
 }
 
+func (repo domainRepo) RetrieveBasicInfoByID(ctx context.Context, id string) (auth.Domain, error) {
```
Why do we need a new repo method, i.e. why don't we use `RetrieveByID`?

WashingtonKK replied:
Because RetrieveByID returns more info than we need. Do you propose we modify the existing method to perform both functions instead?

WashingtonKK replied:
Also, having this dedicated method is self documenting so it's easy to understand it's purpose on the service layer.

rodneyosodo commented on users/service.go:
```diff
@@ -121,11 +121,22 @@ func (svc service) ViewClient(ctx context.Context, token, id string) (mgclients.
+			return mgclients.Client{}, errors.Wrap(svcerr.ErrViewEntity, clientErr)
+		}
+		client.Credentials.Secret = ""
+		return client, err
```
Why are we returning error on successful execution?

rodneyosodo commented on users/postgres/clients.go:
```diff
@@ -123,6 +123,36 @@ func (repo clientRepo) RetrieveByID(ctx context.Context, id string) (mgclients.C
 }
 
+func (repo clientRepo) RetrieveBasicInfoByID(ctx context.Context, id string) (mgclients.Client, error) {
+	q := `SELECT id, name, tags, metadata FROM clients WHERE id = :id`
```
```suggestion
	q := `SELECT id, name FROM clients WHERE id = :id`
```

rodneyosodo replied:
Metadata and tags are not basic information since they can contain user information for example profile picture

rodneyosodo commented on pkg/clients/postgres/clients.go:
```diff
@@ -239,6 +239,32 @@ func (repo *Repository) RetrieveAllBasicInfo(ctx context.Context, pm clients.Pag
 }
 
+func (repo *Repository) RetrieveBasicInfoByID(ctx context.Context, id string) (clients.Client, error) {
+	q := `SELECT c.id, c.name, c.created_at, c.updated_at FROM clients c WHERE id = :id AND status = :status`
```
Are we retrieving only enabled clients or all clients?

rodneyosodo commented on pkg/clients/postgres/clients.go:
```diff
@@ -239,6 +239,32 @@ func (repo *Repository) RetrieveAllBasicInfo(ctx context.Context, pm clients.Pag
 }
 
+func (repo *Repository) RetrieveBasicInfoByID(ctx context.Context, id string) (clients.Client, error) {
+	q := `SELECT c.id, c.name, c.created_at, c.updated_at FROM clients c WHERE id = :id AND status = :status`
```
```suggestion
	q := `SELECT c.id, c.name FROM clients c WHERE id = :id;`
```

rodneyosodo replied:
Also update `RetrieveAllBasicInfo`

rodneyosodo commented on invitations/middleware/tracing.go:
```diff
@@ -24,8 +24,8 @@ func Tracing(svc invitations.Service, tracer trace.Tracer) invitations.Service {
-		attribute.String("user_id", invitation.UserID),
+		attribute.String("domain_id", invitation.Domain.ID),
+		attribute.String("user_id", invitation.User.ID),
 	))
```
Let's enrich the view invitation with name. This applies to logging too

WashingtonKK committed:
Remove couple between invitations and user service

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
Update mocks

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

revert changes on invitations service

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
View basic info for entities

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix testsg

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
Update api docs

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

remove basic information from things and groups

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

add undocumented status code

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

Enhance mocks and fix tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

fix linter and remove debug logs

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

change member relation to permission

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

arvindh123 commented on auth/service.go:
```diff
@@ -560,21 +560,33 @@ func (svc service) CreateDomain(ctx context.Context, token string, d Domain) (do
+	basicInfo.Name = dom.Name
+	basicInfo.Alias = dom.Alias
+	basicInfo.Metadata = dom.Metadata
+
```
Should we expose metadata to public? 
Name and Alias looks sufficent for me.

arvindh123 commented on internal/groups/service.go:
```diff
@@ -112,7 +112,12 @@ func (svc service) CreateGroup(ctx context.Context, token, kind string, g groups
+			return groups.Group{}, errors.Wrap(repoerr.ErrViewEntity, err)
+		}
+
+		return groups.Group{Name: group.Name, Description: group.Description}, nil
```
Why we are exposing the group ? 
Does it required for this PR ?

arvindh123 commented on things/service.go:
```diff
@@ -134,7 +134,11 @@ func (svc service) CreateThings(ctx context.Context, token string, cls ...mgclie
+		if err != nil {
+			return mgclients.Client{}, errors.Wrap(svcerr.ErrViewEntity, err)
+		}
+		return mgclients.Client{Name: client.Name}, nil
```
Same here, 
Do we want to expose things to public ?

arvindh123 reviewed (commented):
Do we need to expose things and groups for public access? 

The related [issue](https://github.com/absmach/magistrala/issues/1987) states only exposing user basic info and domain basic info to authenticated user

arvindh123 commented on users/service.go:
```diff
@@ -115,17 +115,32 @@ func (svc service) RefreshToken(ctx context.Context, refreshToken, domainID stri
+
+		return basicInfo, nil
+	}
 	client, err := svc.clients.RetrieveByID(ctx, id)
```
@WashingtonKK  Suggesting below code for simplification and readable 
```suggestion
	if tokenUserID != id {
		if err := svc.checkSuperAdmin(ctx, tokenUserID); err != nil {
			client, clientErr := svc.clients.RetrieveByID(ctx, id)
			if clientErr != nil {
				return mgclients.Client{}, errors.Wrap(svcerr.ErrViewEntity, clientErr)
			}

			basicInfo.Name = client.Name
			basicInfo.ID = client.ID

			return basicInfo, nil
		}
	}

	client, err := svc.clients.RetrieveByID(ctx, id)
```

arvindh123 commented on auth/service.go:
```diff
@@ -560,21 +560,33 @@ func (svc service) CreateDomain(ctx context.Context, token string, d Domain) (do
+	basicInfo.Alias = dom.Alias
+	basicInfo.Metadata = dom.Metadata
+
+	return basicInfo, errors.Wrap(svcerr.ErrAuthorization, err)
```
We should not return error here

arvindh123 commented on auth/service.go:
```diff
@@ -560,21 +560,33 @@ func (svc service) CreateDomain(ctx context.Context, token string, d Domain) (do
+
+	return basicInfo, errors.Wrap(svcerr.ErrAuthorization, err)
 }
 
```
```suggestion
func (svc service) RetrieveDomain(ctx context.Context, token, id string) (Domain, error) {
	res, err := svc.Identify(ctx, token)
	if err != nil {
		return Domain{}, errors.Wrap(svcerr.ErrAuthentication, err)
	}
	dom, err := svc.domains.RetrieveByID(ctx, id)
	if err != nil {
		return Domain{}, errors.Wrap(svcerr.ErrViewEntity, err)
	}

	if err = svc.Authorize(ctx, PolicyReq{
		Subject:     res.Subject,
		SubjectType: UserType,
		SubjectKind: UsersKind,
		Object:      id,
		ObjectType:  DomainType,
		Permission:  MemberRelation,
	}); err != nil {
		return Domain{Name: dom.Name, Alias: dom.Alias}, nil
	}
	return dom, nil
}

```

arvindh123 reviewed (changes requested):
Please optimize the function and Please test the API response manually for all case

rodneyosodo commented on users/service.go:
```diff
@@ -120,16 +120,22 @@ func (svc service) ViewClient(ctx context.Context, token, id string) (mgclients.
 	if tokenUserID != id {
 		if err := svc.checkSuperAdmin(ctx, tokenUserID); err != nil {
-			return mgclients.Client{}, err
+			client, clientErr := svc.clients.RetrieveByID(ctx, id)
```
`RetrieveByID` is being called twice. No need for this you can refactor it to something like

```go
client, err := svc.clients.RetrieveByID(ctx, id)
if err != nil {
	return mgclients.Client{}, errors.Wrap(svcerr.ErrViewEntity, err)
}

if tokenUserID != id {
	if err := svc.checkSuperAdmin(ctx, tokenUserID); err != nil {
		return mgclients.Client{Name: client.Name, ID: client.ID}, nil
	}
}

client.Credentials.Secret = ""

```

rodneyosodo commented on auth/service.go:
```diff
@@ -560,20 +560,24 @@ func (svc service) CreateDomain(ctx context.Context, token string, d Domain) (do
+		ObjectType:  DomainType,
+		Permission:  MembershipPermission,
+	}); err != nil {
+		return Domain{Name: dom.Name, Alias: dom.Alias}, nil
```
```suggestion
		return Domain{ID: domain.ID, Name: domain.Name, Alias: domain.Alias}, nil
```

rodneyosodo replied:
Add ID

rodneyosodo commented on auth/service.go:
```diff
@@ -560,20 +560,24 @@ func (svc service) CreateDomain(ctx context.Context, token string, d Domain) (do
+	if err != nil {
+		return Domain{}, errors.Wrap(svcerr.ErrAuthentication, err)
 	}
 	dom, err := svc.domains.RetrieveByID(ctx, id)
```
```suggestion
	domain, err := svc.domains.RetrieveByID(ctx, id)
```

WashingtonKK committed:
refactor retrieve domain

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
fix tests and add id to domain

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

rodneyosodo reviewed (dismissed):
LGTM

dborovcanin reviewed (changes requested):
@WashingtonKK Please add API test scenario (fetch user details for admin and non-admin where responses contain all and only basic data, respectively).

WashingtonKK committed:
feat(sdk): add api test cases for changes

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

rodneyosodo commented on pkg/sdk/go/users_test.go:
```diff
@@ -399,6 +404,13 @@ func TestClient(t *testing.T) {
 			err:      nil,
 		},
+		{
+			desc:     "view client successfully as admin",
```
This test it wrong since we use the same token both for admin and non admin

WashingtonKK committed:
fix: use separate admin token in view client test

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
fix: property based tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

rodneyosodo commented on pkg/sdk/go/users_test.go:
```diff
@@ -419,15 +432,23 @@ func TestClient(t *testing.T) {
+		if tc.token == invalidToken {
+			repoCall = auth.On("Identify", mock.Anything, &magistrala.IdentityReq{Token: invalidToken}).Return(&magistrala.IdentityRes{}, svcerr.ErrAuthentication)
 		}
 		repoCall1 := auth.On("Authorize", mock.Anything, mock.Anything).Return(&magistrala.AuthorizeRes{Authorized: true}, nil)
```
All authorize calls pass even if the user is not admin

WashingtonKK committed:
fix: API tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
fix: prevent retrievebyID from returning same response

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

rodneyosodo commented on things/service.go:
```diff
@@ -128,7 +128,7 @@ func (svc service) CreateThings(ctx context.Context, token string, cls ...mgclie
 	_, err := svc.authorize(ctx, "", auth.UserType, auth.TokenKind, token, auth.ViewPermission, auth.ThingType, id)
 	if err != nil {
-		return mgclients.Client{}, err
+		return mgclients.Client{}, errors.Wrap(svcerr.ErrAuthorization, err)
```
Don't wrap it. It is already wrapped

WashingtonKK committed:
fix: remove redundant error wrap

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

dborovcanin commented on users/service.go:
```diff
@@ -133,16 +133,17 @@ func (svc service) ViewClient(ctx context.Context, token, id string) (mgclients.
 		return mgclients.Client{}, err
 	}
 
+	client, err := svc.clients.RetrieveByID(ctx, id)
```
We should not be able to see users' metadata and tags when listing users. I.e. the result of fetching a single user can only be a superset of data returned by fetching users during listing.

WashingtonKK committed:
fix: return basic info when listing members

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
fix: tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
fix: tests

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
refactor service layer

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
fix ci

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

dborovcanin commented on auth/service_test.go:
```diff
@@ -1843,17 +1841,19 @@ func TestRetrieveDomain(t *testing.T) {
-			token:    accessToken,
-			domainID: "",
-			err:      nil,
+			desc:           "retrieve domain with empty domainID",
```
...domain id

dborovcanin commented on users/service.go:
```diff
@@ -187,6 +188,11 @@ func (svc service) ListClients(ctx context.Context, token string, pm mgclients.P
+
+	for i := range pg.Clients {
+		pg.Clients[i] = mgclients.Client{Name: pg.Clients[i].Name, ID: pg.Clients[i].ID}
+	}
```
```go
for i, c := range pg.Clients {
    pg.Clients[i] = mgclients.Client{ID: c.ID, Name: c.Name}
}
```

dborovcanin commented on users/service.go:
```diff
@@ -483,6 +489,10 @@ func (svc service) ListMembers(ctx context.Context, token, objectKind, objectID
 		return mgclients.MembersPage{}, errors.Wrap(svcerr.ErrViewEntity, err)
 	}
 
+	for i := range cp.Clients {
```
Same.

WashingtonKK committed:
fix: refactor code'

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>

WashingtonKK committed:
fix: refactor code'

Signed-off-by: WashingtonKK <washingtonkigan@gmail.com>
